src/nfa/nfa-intersection.cc

```diff
@@ -19,375 +19,191 @@
-            // Check for lhs epsilon transitions.
-            if (lhs_state_transitions.symbol == epsilon) {
-                compute_for_lhs_state_epsilon_transitions(lhs_state_transitions);
-            }
-
-            for (const auto& rhs_state_transitions: rhs.transitionrelation[pair_to_process.second]) {
-                // Find all transitions that have the same symbol for first and the second state in the pair_to_process.
-                if (lhs_state_transitions.symbol == rhs_state_transitions.symbol) {
-                    compute_for_same_symbols(lhs_state_transitions, rhs_state_transitions);
-                }
-            }
-        }
-
-        // Check for rhs epsilon transitions.
-        add_rhs_epsilon_transitions();
-    }
-
-    /**
-     * Check for epsilon transitions in case only rhs has any transitions and add them.
-     */
-    void add_rhs_epsilon_transitions() {
-        for (const auto& rhs_state_transitions: rhs.transitionrelation[pair_to_process.second]) {
-            if (rhs_state_transitions.symbol == epsilon) {
-                compute_for_rhs_state_epsilon_transitions(rhs_state_transitions);
-            }
-        }
-    }
-
-    /**
-     * Initialize pairs to process with initial state pairs.
-     */
-    void initialize_pairs_to_process() {
-        for (const State this_initial_state : lhs.initialstates) {
-            for (const State other_initial_state : rhs.initialstates) {
-                handle_initial_state_pairs(this_initial_state, other_initial_state);
-            }
-        }
-    }
-
-    /**
-     * Add transition to the product.
-     * @param[in] intersection_transition State transitions to add to the product.
-     */
-    void add_product_transition(const TransSymbolStates& intersection_transition)
-    {
-        if (intersection_transition.states_to.empty()) { return; }
-
-        auto& intersect_state_transitions{ product.transitionrelation[product_map[pair_to_process]] };
-        auto symbol_transitions_iter{ intersect_state_transitions.find(intersection_transition) };
-        if (symbol_transitions_iter == intersect_state_transitions.end()) {
-            intersect_state_transitions.push_back(intersection_transition);
-        }
-        else {
-            // Product already has some target states for the given symbol from the current product state.
-            union_to_left(symbol_transitions_iter->states_to, intersection_transition.states_to);
+/**
+ * Create product state and its transitions.
+ * @param[out] product Created product automaton.
+ * @param[out] product_map Created product map.
+ * @param[out] pairs_to_process Set of product states to process
+ * @param[in] lhs_state_to Target state in NFA @c lhs.
+ * @param[in] rhs_state_to Target state in NFA @c rhs.
+ * @param[out] intersect_transitions Transitions of the product state.
+ */
+void create_product_state_and_trans(Nfa& product, ProductMap& product_map, const Nfa& lhs, const Nfa& rhs,
```

Comment:
I'm thinking of maybe rewriting the signatures of the function from:
```
void fun(par1, par2, par3
              par4, par5)
```
To something like:
```
void fun(
    par1,
    par2,
    ...
    parn
):
```

Reply:
The number of parameters for these functions grew a bit due to having to pass the original class member variables. I will refactor this as you suggest.